cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/WorkflowNodeCodec.java

```diff
@@ -0,0 +1,120 @@
+
+package co.cask.cdap.internal.app;
+
+import co.cask.cdap.api.schedule.SchedulableProgramType;
+import co.cask.cdap.api.workflow.ScheduleProgramInfo;
+import co.cask.cdap.api.workflow.WorkflowActionNode;
+import co.cask.cdap.api.workflow.WorkflowActionSpecification;
+import co.cask.cdap.api.workflow.WorkflowForkNode;
+import co.cask.cdap.api.workflow.WorkflowNode;
+import co.cask.cdap.api.workflow.WorkflowNodeType;
+import com.google.common.reflect.TypeToken;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+
+import java.lang.reflect.Type;
+import java.util.List;
+
+/**
+ * Codec to serialize and deserialize {@link WorkflowNode}
+ */
+final class WorkflowNodeCodec extends AbstractSpecificationCodec<WorkflowNode> {
+  @Override
+  public JsonElement serialize(WorkflowNode src, Type typeOfSrc, JsonSerializationContext context) {
+    JsonObject jsonObj = new JsonObject();
+    jsonObj.add("nodeId", new JsonPrimitive(src.getNodeId()));
+    jsonObj.add("nodeType", new JsonPrimitive(src.getType().toString()));
+
+    switch (src.getType()) {
+      case ACTION:
+        serializeActionNode(src, jsonObj, context);
+        break;
+      case FORK:
+        serializeForkNode(src, jsonObj, context);
+        break;
+      case CONDITION:
+        // no-op
+        break;
+      default:
+        // no-ops
+        break;
+    }
+    return jsonObj;
+  }
+
+  private void serializeActionNode(WorkflowNode node, JsonObject jsonObj, JsonSerializationContext context) {
+    WorkflowActionNode actionNode = (WorkflowActionNode) node;
+    jsonObj.add("program", context.serialize(actionNode.getProgram(), ScheduleProgramInfo.class));
+    if (actionNode.getProgram().getProgramType() == SchedulableProgramType.CUSTOM_ACTION) {
+      jsonObj.add("actionSpecification", context.serialize(actionNode.getActionSpecification(),
+                                                           WorkflowActionSpecification.class));
+    }
+  }
+
+  private void serializeForkNode(WorkflowNode node, JsonObject jsonObj, JsonSerializationContext context) {
+    WorkflowForkNode forkNode = (WorkflowForkNode) node;
+    Type type = new TypeToken<List<List<WorkflowNode>>>() { }.getType();
+    jsonObj.add("branches", context.serialize(forkNode.getBranches(), type));
+  }
+
+  @Override
+  public WorkflowNode deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+    throws JsonParseException {
+    JsonObject jsonObj = json.getAsJsonObject();
+
+    String nodeId = context.deserialize(jsonObj.get("nodeId"), String.class);
+    WorkflowNodeType type = context.deserialize(jsonObj.get("nodeType"), WorkflowNodeType.class);
+
+    WorkflowNode node = null;
+    switch (type) {
+      case ACTION:
+        node = createActionNode(nodeId, jsonObj, context);
```

Comment:
No need to have separate methods. After you know the `type` (not even need the `nodeId`), do:

```java
switch (type) {
  case ACTION:
    return context.deserialize(json, WorkflowActionNode.class);
  case FORK:
    return context.deserialize(json, WorkflowForkNode.class);
}
```